### PR TITLE
[Serializer] Add a MaxDepth option

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/issues/14924#issuecomment-111977173, https://github.com/dunglas/DunglasApiBundle/issues/104#issuecomment-110644569 |
| License | MIT |
| Doc PR | todo |

Add a max depth option during the normalization process. Especially useful when normalizing trees.

Usage:

``` php
use Symfony\Component\Serializer\Annotation\MaxDepth;

class MyObj
{
    /**
     * @MaxDepth(2)
     */
    public $foo;

    /**
     * @var self
     */
    public $child;
}

use Doctrine\Common\Annotations\AnnotationReader;
use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
use Symfony\Component\Serializer\Serializer;
use Symfony\Component\Serializer\Mapping\Loader\AnnotationLoader;
use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactory;

$classMetadataFactory = new ClassMetadataFactory(new AnnotationLoader(new AnnotationReader()));
$normalizer = new ObjectNormalizer($classMetadataFactory);
$serializer = new Serializer(array($this->normalizer));

$level1 = new MyObj();
$level1->foo = 'level1';

$level2 = new MyObj();
$level2->foo = 'level2';
$level1->child = $level2;

$level3 = new MyObj();
$level3->foo = 'level3';
$level2->child = $level3;
$result = $serializer->normalize($level1, null, array(ObjectNormalizer::ENABLE_MAX_DEPTH => true));
/*      $result = array(
            'foo' => 'level1',
            'child' => array(
                    'foo' => 'level2',
                    'child' => array(
                            'child' => null,
                        ),
                ),
        );
*/
```
- [x] Metadata support
- [x] `@MaxDepth(2)` annotation and loader
- [x] XML loader
- [x] YAML loader
- [x] Delegate recursive normalization at the end of the process
- [x] Use constants with Late Static Binding in the abstract class instead of raw strings
- [x] Move common logic to the abstract class

/cc @mRoca @csarrazi

### CODE REVIEW

File: src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php [hunk 1]
```diff
@@ -42,6 +42,20 @@ public function addGroup($group);
     public function getGroups();
 
     /**
+     * Sets the serialization max depth for this attribute.
+     *
+     * @param int|null $maxDepth
+     */
+    public function setMaxDepth($maxDepth);
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php]: Isn't that a BC break?

File: src/Symfony/Component/Serializer/Annotation/MaxDepth.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+class MaxDepth
+{
+    /**
+     * @var int
+     */
+    private $maxDepth;
+
+    public function __construct(array $data)
+    {
+        if (!isset($data['value']) || !$data['value']) {
```
Comment (jderusse) [inline src/Symfony/Component/Serializer/Annotation/MaxDepth.php]: What's about `if (empty($data['value'])) {`

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 3]
```diff
@@ -55,42 +56,59 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
...
-                    if ($groups instanceof Groups) {
-                        foreach ($groups->getGroups() as $group) {
+                foreach ($this->reader->getPropertyAnnotations($property) as $annotation) {
+                    if ($annotation instanceof Groups) {
+                        foreach ($annotation->getGroups() as $group) {
                             $attributesMetadata[$property->name]->addGroup($group);
                         }
                     }
 
+                    if ($annotation instanceof MaxDepth) {
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Why not using an ``elseif`` here ?

File: src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php]: Good catch, needs a new interface...

Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php]: By the way, I think than introducing such interfaces was a (my) mistake. Serializer Metadata are value objects. Nobody will (nor should) implement these interfaces. Composition is better for "extending" metadata.

What do you think about marking them as `@internal` (and even deprecated) @symfony/deciders?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 4]
```diff
@@ -287,4 +296,116 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+            return true;
+        }
+
+        if ($keyExist) {
+            ++$context[$key];
+        } else {
+            $context[$key] = 1;
+        }
+
+        if ($attribute === 'foo') {
```
Comment (jderusse) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: ?

File: src/Symfony/Component/Serializer/Annotation/MaxDepth.php (hunk 2 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Annotation/MaxDepth.php]: :+1:

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 3 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: changed

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Oops ^^

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 3 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Not at this line ;-)

File: src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php]: #17114

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 5]
```diff
@@ -287,4 +294,108 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+    /**
+     * Sets an attribute and apply the name converter if necessary.
+     *
+     * @param array  $data
+     * @param string $attribute
+     * @param mixed  $attributeValue
+     *
+     * @return array
+     */
+    protected function setAttribute($data, $attribute, $attributeValue)
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: What do you think about using the reference operator ``&$data`` to avoid useless copy of the array?

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: You can also add an ``array`` type hint.

**Comment by Ener-Getick** [comment]
:+1: to merge this.
Just needs to take a decision about the interface.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 5 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I prefer to avoid references when possible (http://schlueters.de/blog/archives/125-do-not-use-php-references.html).

Typehint added.

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Ok makes sense.

**Comment by Ener-Getick** [comment]
Sounds good to me
Status: Reviewed

File: src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/Serializer/Mapping/AttributeMetadataInterface.php]: #17114 merged

**Comment by dunglas** [comment]
ping @symfony/deciders 

Would be nice to merge soon to be able to work on #16143 without creating conflicts.

File: src/Symfony/Component/Serializer/Annotation/MaxDepth.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    public function __construct(array $data)
+    {
+        if (empty($data['value'])) {
+            throw new InvalidArgumentException(sprintf('Parameter of annotation "%s" cannot be empty.', get_class($this)));
+        }
+
+        if (!is_int($data['value'])) {
+            throw new InvalidArgumentException(sprintf('Parameter of annotation "%s" must be an int.', get_class($this)));
+        }
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Annotation/MaxDepth.php]: What about also checking that the value is greater than zero?

File: src/Symfony/Component/Serializer/Mapping/ClassMetadata.php [hunk 7]
```diff
@@ -28,7 +28,7 @@ class ClassMetadata implements ClassMetadataInterface
     public $name;
 
     /**
-     * @var AttributeMetadataInterface[]
+     * @var array
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/ClassMetadata.php]: Why this change?

Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/ClassMetadata.php]: Because the key is a `string` and some tools like Scrutinizr consider that the `[]` notation is only applicable for packed arrays (consecutive `int` as keys).

File: src/Symfony/Component/Serializer/Annotation/MaxDepth.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    private $maxDepth;
+
+    public function __construct(array $data)
+    {
+        if (empty($data['value'])) {
+            throw new InvalidArgumentException(sprintf('Parameter of annotation "%s" cannot be empty.', get_class($this)));
+        }
+
+        if (!is_int($data['value'])) {
+            throw new InvalidArgumentException(sprintf('Parameter of annotation "%s" must be an int.', get_class($this)));
```
Comment (javiereguiluz) [inline src/Symfony/Component/Serializer/Annotation/MaxDepth.php]: `must be an int` -> `must be a integer`  (or even `must be a positive integer` if we follow @xabbuh's comments about this)

File: src/Symfony/Component/Serializer/Mapping/ClassMetadata.php (hunk 7 above)
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/ClassMetadata.php]: I see, so technically this is a bugfix, isn't it? I would say let's improve the PHPdoc in all versions where this applies then.

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 9]
```diff
@@ -68,23 +68,17 @@ public function normalize($object, $format = null, array $context = array())
...
-                    throw new LogicException(sprintf('Cannot normalize attribute "%s" because injected serializer is not a normalizer', $property->name));
-                }
-
-                $attributeValue = $this->serializer->normalize($attributeValue, $format, $context);
-            }
 
-            $propertyName = $property->name;
-            if ($this->nameConverter) {
-                $propertyName = $this->nameConverter->normalize($propertyName);
+            // Recursive call are done at the end of the process to allow @MaxDepth to work
```
Comment (javiereguiluz) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: `Recursive call are done` -> `Recursive call is done` or `Recursive calls are done`

File: src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php [hunk 10]
```diff
@@ -73,22 +73,16 @@ public function normalize($object, $format = null, array $context = array())
                 $attributeValue = call_user_func($this->callbacks[$attribute], $attributeValue);
             }
 
+            // Recursive call are done at the end of the process to allow @MaxDepth to work
```
Comment (javiereguiluz) [inline src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php]: `Recursive call are done` -> `Recursive call is done` or `Recursive calls are done`

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 11]
```diff
@@ -68,29 +71,40 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
...
-                                $attributeMetadata->addGroup($group);
-                            }
-                        } else {
-                            throw new MappingException(sprintf('Groups on "%s::%s" cannot be added. Groups can only be added on methods beginning with "get", "is", "has" or "set".', $className, $method->name));
-                        }
+            if ($method->getDeclaringClass()->name !== $className) {
+                continue;
+            }
+
+            $accessorOrMutator = preg_match('/^(get|is|has|set)(.+)$/i', $method->name, $matches);
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I would name the variable `$isAccessorOrMutator` to reflect the boolean value.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 12]
```diff
@@ -68,29 +71,40 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
...
+            if ($accessorOrMutator) {
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I know we are not that strict with the return value of `preg_match()` in other place, but I would prefer `false !== $accessorOrMutator`.

File: src/Symfony/Component/Serializer/Tests/Annotation/MaxDepthTest.php [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class MaxDepthTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @expectedException \Symfony\Component\Serializer\Exception\InvalidArgumentException
+     */
+    public function testNotAnIntMaxDepthParameter()
+    {
+        new MaxDepth(array('value' => 'coopTilleuls'));
```
Comment (javiereguiluz) [inline src/Symfony/Component/Serializer/Tests/Annotation/MaxDepthTest.php]: I know I've told you this before, but using the name of a real company as the value of the tests looks wrong to me and is inconsistent with the rest of Symfony codebase.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 12 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: It doesn't work. `preg_match ` returns `0` if the pattern is not found and `false` in case of error.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 11 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: We do that for methods but does it make sense for variables? We doesn't name them `$intFoo` or `$stringBar`.

File: src/Symfony/Component/Serializer/Mapping/ClassMetadata.php (hunk 7 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/ClassMetadata.php]: Not really a bug (it just improves static analysis) but it makes sense to apply this on every versions.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 12 above)
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Ah yeah, you're right.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 11 above)
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I think prefixing a variable with `is` is a bit different than prefixing it with an actual type. But I wouldn't insist on this if you don't like it.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 12 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I've no strong opinion about that. What do we use in usually?

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 11 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I've no strong opinion about that. What do we use usually?

File: src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd [hunk 14]
```diff
@@ -47,10 +47,11 @@
         Contains serialization groups for a attributes. The name of the attribute should be given in the "name" option.
       ]]></xsd:documentation>
         </xsd:annotation>
-        <xsd:sequence>
+        <xsd:sequence minOccurs="0">
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd]: looks like this should also be done on older branches

File: src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php [hunk 15]
```diff
@@ -75,9 +76,13 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
 
                     if (isset($data['groups'])) {
                         foreach ($data['groups'] as $group) {
-                            $attributeMetadata->addGroup($group);
+                            $attributeMetadata->addGroup((string) $group);
                         }
                     }
+
+                    if (isset($data['max_depth'])) {
+                        $attributeMetadata->setMaxDepth((int) $data['max_depth']);
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php]: We should check the value (we do that for the annotation and we have an implicit check in the XML loader due to the XSD file).

File: src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php [hunk 16]
```diff
@@ -65,6 +65,7 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
 
             if (isset($yaml['attributes']) && is_array($yaml['attributes'])) {
                 $attributesMetadata = $classMetadata->getAttributesMetadata();
+
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php]: Doing this everywhere (on all versions)?

File: src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php [hunk 17]
```diff
@@ -75,9 +76,13 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
 
                     if (isset($data['groups'])) {
                         foreach ($data['groups'] as $group) {
-                            $attributeMetadata->addGroup($group);
+                            $attributeMetadata->addGroup((string) $group);
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php]: same here

File: src/Symfony/Component/Serializer/Mapping/Loader/XmlFileLoader.php [hunk 18]
```diff
@@ -62,6 +62,10 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata)
                 foreach ($attribute->group as $group) {
                     $attributeMetadata->addGroup((string) $group);
                 }
+
+                if (isset($attribute['max-depth'])) {
+                    $attributeMetadata->setMaxDepth((int) $attribute['max-depth']);
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/XmlFileLoader.php]: Check for a value to be greater than zero?

File: src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd (hunk 14 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd]: IMO it's not necessary: previously metadata was only used for groups. It makes sense to require it.
But with this new annotation (only in 3.1+) you can use metadata for groups or max depth. It makes sense to make groups optional only for 3.1+.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 19]
```diff
@@ -25,6 +26,12 @@
  */
 abstract class AbstractNormalizer extends SerializerAwareNormalizer implements NormalizerInterface, DenormalizerInterface
 {
+    const CIRCULAR_REFERENCE_LIMIT = 'circular_reference_limit';
+    const OBJECT_TO_POPULATE = 'object_to_populate';
+    const GROUPS = 'groups';
+    const ENABLE_MAX_DEPTH = 'enable_max_depth';
+    const DEPTH_KEY_PATTERN = 'depth_%s::%s';
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I would introduce the constants in older versions too.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 20]
```diff
@@ -287,4 +294,108 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+     *
+     * @return bool
+     */
+    protected function isAttributeToNormalize($object, $attributeName, &$context)
+    {
+        return !in_array($attributeName, $this->ignoredAttributes) && !$this->isMaxDepthReached(get_class($object), $attributeName, $context);
+    }
+
+    /**
+     * Sets an attribute and apply the name converter if necessary.
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: applies

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 21]
```diff
@@ -287,4 +294,108 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+     *
+     * @param array  $data
+     * @param string $attribute
+     * @param mixed  $attributeValue
+     *
+     * @return array
+     */
+    protected function setAttribute(array $data, $attribute, $attributeValue)
+    {
+        if ($this->nameConverter) {
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: `null !== $this->nameConverter`

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 22]
```diff
@@ -287,4 +294,108 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+     *
+     * @param string $class
+     * @param string $attribute
+     * @param array  $context
+     *
+     * @return bool
+     */
+    private function isMaxDepthReached($class, $attribute, array &$context)
+    {
+        if (!$this->classMetadataFactory || !isset($context[static::ENABLE_MAX_DEPTH])) {
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: `null === $this->classMetadataFactory [...]`

File: src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd (hunk 14 above)
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd]: I see, but then we should update the inline comments which both still refer to groups only.

File: src/Symfony/Component/Serializer/Mapping/ClassMetadata.php (hunk 7 above)
Comment (csarrazi) [inline src/Symfony/Component/Serializer/Mapping/ClassMetadata.php]: You could use `array|AttributeMetadataInterface[]`, no?

File: src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd [hunk 23]
```diff
@@ -47,10 +47,11 @@
...
             <xsd:element name="group" type="xsd:string" maxOccurs="unbounded" />
         </xsd:sequence>
         <xsd:attribute name="name" type="xsd:string" use="required" />
+        <xsd:attribute name="max-depth" type="xsd:int" />
```
Comment (csarrazi) [inline src/Symfony/Component/Serializer/Mapping/Loader/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd]:
```xml
<xsd:attribute name="max-depth">
    <xsd:simpleType>
        <xsd:restriction base="xs:integer">
            <xsd:minInclusive value="0"/>
        </xsd:restriction>
    </xsd:simpleType>
</xsd:attribute>
```

We want the restriction to be part of the schema, too.

**Comment by xabbuh** [comment]
I wonder if we could move some more code from the different normalizer implementations into the `AbstractNormalizer` class. They look quite similar.

**Comment by dunglas** [comment]
Status: needs work

File: src/Symfony/Component/Serializer/Mapping/Loader/XmlFileLoader.php (hunk 18 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/XmlFileLoader.php]: I've added the constraint in the XSD instead.

**Comment by dunglas** [comment]
Status: needs review

ping @symfony/deciders

File: src/Symfony/Component/Serializer/Mapping/ClassMetadataInterface.php [hunk 24]
```diff
@@ -41,7 +41,7 @@ public function addAttributeMetadata(AttributeMetadataInterface $attributeMetada
     /**
      * Gets the list of {@link AttributeMetadataInterface}.
      *
-     * @return AttributeMetadataInterface[]
+     * @return @return AttributeMetadataInterface[]
```
Comment (fabpot) [inline src/Symfony/Component/Serializer/Mapping/ClassMetadataInterface.php]: typo

File: src/Symfony/Component/Serializer/Annotation/Groups.php [hunk 25]
```diff
@@ -35,7 +35,7 @@ class Groups
      */
     public function __construct(array $data)
     {
-        if (!isset($data['value']) || !$data['value']) {
+        if (empty($data['value'])) {
```
Comment (fabpot) [inline src/Symfony/Component/Serializer/Annotation/Groups.php]: I prefer the previous explicit condition

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 11 above)
Comment (fabpot) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: $isXXX sound better to me.

**Comment by fabpot** [comment]
Made some small comments, :+1:

File: src/Symfony/Component/Serializer/Annotation/Groups.php (hunk 25 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Annotation/Groups.php]: Me too but I've been asked to change such statements to `empty` on several PR.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Annotation/Groups.php]: Even on this one actually: https://github.com/symfony/symfony/pull/17113#discussion-diff-48291261

Comment (fabpot) [inline src/Symfony/Component/Serializer/Annotation/Groups.php]: That's a shame. We never use empty in Symfony.

File: src/Symfony/Component/Serializer/Annotation/MaxDepth.php (hunk 2 above)
Comment (fabpot) [inline src/Symfony/Component/Serializer/Annotation/MaxDepth.php]: Please, don't use empty.

**Comment by dunglas** [comment]
Should be ok now.

**Comment by fabpot** [comment]
Thank you @dunglas.
